stl/src/atomic_wait.cpp

```diff
@@ -0,0 +1,255 @@
+        switch (_Wait_context & _Atomic_wait_phase_mask) {
+        case _Atomic_wait_phase_wait_none: {
+            _Wait_context = _Atomic_wait_phase_wait_locked;
+            auto& entry   = _Atomic_wait_table_entry(_Storage);
+            ::AcquireSRWLockExclusive(&entry._Lock);
+            [[fallthrough]];
+        }
+
+        case _Atomic_wait_phase_wait_locked: {
+            auto& entry = _Atomic_wait_table_entry(_Storage);
+            ::SleepConditionVariableSRW(&entry._Condition, &entry._Lock, INFINITE, 0);
+            // re-check, and still in _Atomic_wait_phase_wait_locked
+            return;
+        }
+        }
+    }
+
+    void _Atomic_unwait_fallback(const void* const _Storage, unsigned long long& _Wait_context) noexcept {
+        if (_Wait_context & _Atomic_wait_phase_wait_locked) {
+            auto& entry = _Atomic_wait_table_entry(_Storage);
+            ::ReleaseSRWLockExclusive(&entry._Lock);
+        }
+    }
+
+    void _Atomic_notify_fallback(const void* const _Storage) noexcept {
+        auto& entry = _Atomic_wait_table_entry(_Storage);
+        ::AcquireSRWLockExclusive(&entry._Lock);
+        ::ReleaseSRWLockExclusive(&entry._Lock);
+        ::WakeAllConditionVariable(&entry._Condition);
+    }
+
+    struct _Wait_on_address_functions {
+        std::atomic<decltype(&::WaitOnAddress)> _Pfn_WaitOnAddress;
+        std::atomic<decltype(&::WakeByAddressSingle)> _Pfn_WakeByAddressSingle;
+        std::atomic<decltype(&::WakeByAddressAll)> _Pfn_WakeByAddressAll;
+        std::atomic<bool> _Initialized;
+    };
+
+    const _Wait_on_address_functions& _Get_wait_functions() {
+        static _Wait_on_address_functions functions;
+        if (!functions._Initialized.load(std::memory_order_relaxed)) {
+            HMODULE sync_api_module        = ::GetModuleHandle(TEXT("API-MS-WIN-CORE-SYNCH-L1-2-0.DLL"));
+            FARPROC wait_on_address        = ::GetProcAddress(sync_api_module, "WaitOnAddress");
+            FARPROC wake_by_address_single = ::GetProcAddress(sync_api_module, "WakeByAddressSingle");
+            FARPROC wake_by_address_all    = ::GetProcAddress(sync_api_module, "WakeByAddressAll");
+
+            if (wait_on_address != nullptr && wake_by_address_single != nullptr && wake_by_address_all != nullptr) {
+                functions._Pfn_WaitOnAddress.store(
+                    reinterpret_cast<decltype(&::WaitOnAddress)>(wait_on_address), std::memory_order_relaxed);
+                functions._Pfn_WakeByAddressSingle.store(
+                    reinterpret_cast<decltype(&::WakeByAddressSingle)>(wake_by_address_single),
+                    std::memory_order_relaxed);
+                functions._Pfn_WakeByAddressAll.store(
+                    reinterpret_cast<decltype(&::WakeByAddressAll)>(wake_by_address_all), std::memory_order_relaxed);
+            }
+            std::atomic_thread_fence(std::memory_order_seq_cst);
+            functions._Initialized.store(true, std::memory_order_relaxed);

```

Comment:
I'm pretty sure 116 needs to be a load acquire and 132 needs to be store-release to prevent the stores 123-129 from being observed reordered after this store from other threads. (As far as I know `atomic_thread_fence` doesn't help you prevent that since the case there is being observed from other threads but I'm far from a memory model expert)

Reply:
I've tried that questionable thing with fence because I wanted to avoid overhead for `acquire`.
But hey, this code will execute only on x86 / x64, since everything else will have at least Win8.
For x86 / x64 load acquire is just plain load.